tutorials/nemo-retriever-synthetic-data-generation/retriever_hardnegative_miner.py

```diff
@@ -0,0 +1,304 @@
+
+from nemo_curator import ClusteringModel
+from nemo_curator.datasets import DocumentDataset
+from nemo_curator.utils.distributed_utils import NoWorkerError, load_object_on_worker
+
+config = importlib.import_module(
+    "tutorials.nemo-retriever-synthetic-data-generation.config.config"
+)
+RetrieverHardNegativeMiningConfig = config.RetrieverHardNegativeMiningConfig
+
+
+def create_nim_client(base_url, api_key):
+    openai_client = OpenAI(base_url=base_url, api_key=api_key)
+    return openai_client
+
+
+def create_hf_model(model_name_or_path):
+    return SentenceTransformer(model_name_or_path, trust_remote_code=True)
+
+
+class HardNegativeMiner:
+
+    def __init__(
+        self,
+        cfg: RetrieverHardNegativeMiningConfig,
+    ):
+
+        self.model_name = cfg.model_name
+        self.model_type = cfg.model_type
+        self.base_url = cfg.base_url
+        self.api_key = cfg.api_key
+        self.truncate = cfg.truncate
+        self.n_hard_negatives = cfg.hard_negatives_to_mine
+
+        if cfg.passage_prefix:
+            self.passage_prefix = cfg.passage_prefix
+        if cfg.query_prefix:
+            self.query_prefix = cfg.query_prefix
+        if cfg.hard_neg_mining_algorithm:
+            self.hard_neg_mining_algorithm = cfg.hard_neg_mining_algorithm
+        else:
+            print(
+                "hard negative mining algorithm not mentioned in config, using default"
+            )
+            self.hard_neg_mining_algorithm = "topk_percpos"
+        if self.hard_neg_mining_algorithm == "topk_percpos":
+            if cfg.percpos:
+                self.percpos = cfg.percpos
+            else:
+                self.percpos = 0.95
+        elif self.hard_neg_mining_algorithm == "topk_abs":
+            if cfg.max_hardness_threshold:
+                self.max_neg_score_threshold = cfg.max_hardness_threshold
+            else:
+                raise ValueError("Hard negative threshold is required!")
+            if cfg.min_hardness_threshold:
+                self.min_neg_score_threshold = cfg.min_hardness_threshold
+            else:
+                self.min_neg_score_threshold = 0.0
+        if cfg.min_cluster_size:
+            self.min_cluster_size = cfg.min_cluster_size
+        if cfg.max_number_clusters:
+            self.max_number_clusters = cfg.max_number_clusters
+        if cfg.cluster_output_dir:
+            self.cluster_output_dir = cfg.cluster_output_dir
+        if cfg.logger_output_dir:
+            self.logger_output_dir = cfg.logger_output_dir
+
+    def __dask_tokenize__(self):
+        return normalize_token(HardNegativeMiner)
+
+    def assign_ids(self, partition):
+        return partition.assign(doc_id=np.arange(len(partition)) + partition.index[0])
+
+    def repartition_semantic_similarity(
+        self, dataset: DocumentDataset
+    ) -> DocumentDataset:
+        df = dataset.df
+        n_data = df.compute().shape[0]  # number of row items
```

Comment:
This will make sure we dont do double computation here and only pull in shape 

```suggestion
        df = df.persist()
        n_data = df.shape[0].compute()
```